### PR TITLE
Fix clamp(None, nan) to propagate scalar NaNs

### PR DESCRIPTION
**Pull request by sirohikartik**
Fixes #172067
Summary

This PR fixes eager torch.clamp behavior when a scalar NaN is provided as the sole bound (e.g. torch.clamp(x, None, nan)), which previously returned the input unchanged.

Details

Aligned ATen eager semantics with _refs and existing clamp_min / clamp_max scalar-NaN handling.

Added a regression test covering scalar NaN bounds to prevent future regressions.

### CODE REVIEW

File: torch/_refs/__init__.py [hunk 1]
```diff
@@ -2000,10 +2000,19 @@ def clamp(
     min: Optional[TensorOrNumberLikeType] = None,
     max: Optional[TensorOrNumberLikeType] = None,
 ) -> TensorLikeType:
+
+    def _is_scalar_nan(x:object) -> bool:
```
Comment (Skylion007) [inline torch/_refs/__init__.py]: This should be a TypeGuard[float|complex] return type

File: aten/src/ATen/native/TensorCompare.cpp [hunk 2]
```diff
@@ -836,9 +837,29 @@ TORCH_IMPL_FUNC(clamp_out)
...
+      // (aligns clamp(None, NaN) with clamp_max/min and _refs behavior)
+      at::fill_(
+        const_cast<Tensor&>(result),
+        std::numeric_limits<double>::quiet_NaN()
+      );
+    }
+    else{
     clamp_min_scalar_stub(device_type(), *this, min.get());
+    }
   }
```
Comment (lakshayg) [inline aten/src/ATen/native/TensorCompare.cpp:846]: Something like this would be cleaner IMO
```cpp
// If either of the bounds are `NaN`, the output should be `NaN`
bool min_is_nan = min.has_value() && (min.get().toDouble() != min.get().toDouble());
bool max_is_nan = max.has_value() && (max.get().toDouble() != max.get().toDouble());

if (min_is_nan || max_is_nan) {
  at::fill_(
    const_cast<Tensor&>(result),
    std::numeric_limits<double>::quiet_NaN());
} else if (min && max) {
  clamp_scalar_stub(device_type(), *this, min.get(), max.get());
} else if (max) {
  clamp_max_scalar_stub(device_type(), *this, max.get());
} else if (min) {
  clamp_min_scalar_stub(device_type(), *this, min.get());
}
```

File: torch/_refs/__init__.py [hunk 3]
```diff
@@ -2000,10 +2000,19 @@ def clamp(
...
+        return isinstance(x, (float, complex)) and math.isnan(x)
     # NOTE: grad behavior with implementation `where` is not consistent on `nan`
     if min is None and max is None:
         msg = "clamp called but both min and max are none!"
         raise ValueError(msg)
+    if min is None and _is_scalar_nan(max):
+        return torch.full_like(a, max)
```
Comment (ngimel) [inline torch/_refs/__init__.py]: this would break the gradient chain, and also decompositions below are already supposed to propagate nans correctly

File: torch/_refs/__init__.py (hunk 1 above)
Comment (sirohikartik) [reply torch/_refs/__init__.py]: Thanks for pointing this out. I’ve removed the `_is_scalar_nan` helper entirely from `_refs` and reverted to the existing `where`-based logic to preserve autograd semantics.

File: aten/src/ATen/native/TensorCompare.cpp (hunk 2 above)
Comment (sirohikartik) [reply aten/src/ATen/native/TensorCompare.cpp:846]: Good suggestion, thanks! I’ve refactored the NaN handling into a single `min_is_nan / max_is_nan` check as proposed to simplify the control flow.

File: torch/_refs/__init__.py (hunk 3 above)
Comment (sirohikartik) [reply torch/_refs/__init__.py]: Thanks for the catch, you’re right. I’ve removed the scalar-NaN early returns from `_refs` and now rely on the existing decomposition logic to propagate NaNs without breaking the gradient chain.

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
@claude Although you said the diff is good, it is kind of pointless right? Like, why would you pass in a NaN scalar. At least the diff doesn't make things slower (except maybe a small constant amount?)
